File: scripts/check_ll1.py
#!/usr/bin/env python3
"""
LL(1) grammar checker for Concrete.

Parses grammar/concrete.ebnf and verifies the LL(1) property:
for every alternation (A | B | C), the FIRST sets of A, B, C
must be pairwise disjoint (after accounting for nullable rules).

Usage:
    python3 scripts/check_ll1.py [grammar/concrete.ebnf]

Exit code 0 = LL(1) clean, 1 = conflicts found.
"""

import re
import sys
from collections import defaultdict
from pathlib import Path


# ── EBNF Parser ──────────────────────────────────────────────────

class Terminal:
    """A terminal symbol like 'fn' or INT."""
    def __init__(self, name):
        self.name = name
    def __repr__(self):
        return self.name
    def __eq__(self, other):
        return isinstance(other, Terminal) and self.name == other.name
    def __hash__(self):
        return hash(self.name)

class Nonterminal:
    """A nonterminal reference like expr or type."""
    def __init__(self, name):
        self.name = name
    def __repr__(self):
        return f"<{self.name}>"
    def __eq__(self, other):
        return isinstance(other, Nonterminal) and self.name == other.name
    def __hash__(self):
        return hash(("NT", self.name))

class Sequence:
    """A sequence of symbols: A B C."""
    def __init__(self, items):
        self.items = items
    def __repr__(self):
        return " ".join(str(i) for i in self.items)

class Alternation:
    """A choice: A | B | C."""
    def __init__(self, alts):
        self.alts = alts
    def __repr__(self):
        return " | ".join(str(a) for a in self.alts)

class Optional:
    """Zero or one: [ A ]."""
    def __init__(self, body):
        self.body = body
    def __repr__(self):
        return f"[ {self.body} ]"

class Repetition:
    """Zero or more: { A }."""
    def __init__(self, body):
        self.body = body
    def __repr__(self):
        return f"{{ {self.body} }}"

# Token classes for the EBNF parser itself
EPSILON = Terminal("ε")

# Known terminal token classes (not quoted)
TOKEN_CLASSES = {"IDENT", "INT", "FLOAT", "STRING", "CHAR", "LABEL"}


def tokenize_ebnf(text):
    """Tokenize EBNF text into a list of tokens."""
    tokens = []
    i = 0
    while i < len(text):
        # Skip whitespace
        if text[i] in " \t\n\r":
            i += 1
            continue
        # Skip comments (* ... *)
        if text[i:i+2] == "(*":
            end = text.find("*)", i + 2)
            if end == -1:
                raise ValueError(f"Unterminated comment at position {i}")
            i = end + 2
            continue
        # Single-char delimiters
        if text[i] in "=;|[]{}().,+":
            tokens.append(text[i])
            i += 1
            continue
        # Quoted terminal: 'xxx'
        if text[i] == "'":
            j = text.index("'", i + 1)
            tokens.append(("TERM", text[i+1:j]))
            i = j + 1
            continue
        # Identifier (nonterminal or token class)
        if text[i].isalpha() or text[i] == "_":
            j = i
            while j < len(text) and (text[j].isalnum() or text[j] == "_"):
                j += 1
            tokens.append(("ID", text[i:j]))
            i = j
            continue
        raise ValueError(f"Unexpected character {text[i]!r} at position {i}")
    return tokens


def parse_ebnf(text):
    """Parse EBNF text into a dict of {name: body} rules."""
    tokens = tokenize_ebnf(text)
    rules = {}
    i = 0

    def peek():
        nonlocal i
        return tokens[i] if i < len(tokens) else None

    def advance():
        nonlocal i
        t = tokens[i]
        i += 1
        return t

    def expect(val):
        t = advance()
        if t != val:
            raise ValueError(f"Expected {val!r}, got {t!r} at token {i-1}")

    def parse_alternation():
        alts = [parse_sequence()]
        while peek() == "|":
            advance()
            alts.append(parse_sequence())
        if len(alts) == 1:
            return alts[0]
        return Alternation(alts)

    def parse_sequence():
        items = []
        while True:
            p = peek()
            if p is None or p in ("|", ";", "]", "}", ")"):
                break
            items.append(parse_atom())
        if len(items) == 0:
            return Sequence([EPSILON])
        if len(items) == 1:
            return items[0]
        return Sequence(items)

    def parse_atom():
        p = peek()
        if p == "[":
            advance()
            body = parse_alternation()
            expect("]")
            return Optional(body)
        if p == "{":
            advance()
            body = parse_alternation()
            expect("}")
            return Repetition(body)
        if p == "(":
            advance()
            body = parse_alternation()
            expect(")")
            return body
        if isinstance(p, tuple):
            advance()
            kind, val = p
            if kind == "TERM":
                return Terminal(f"'{val}'")
            if kind == "ID":
                if val in TOKEN_CLASSES:
                    return Terminal(val)
                return Nonterminal(val)
        raise ValueError(f"Unexpected token {p!r} at position {i}")

    while i < len(tokens):
        name_tok = advance()
        if not isinstance(name_tok, tuple) or name_tok[0] != "ID":
            raise ValueError(f"Expected rule name, got {name_tok!r}")
        name = name_tok[1]
        expect("=")
        body = parse_alternation()
        expect(";")
        rules[name] = body

    return rules


# ── FIRST / FOLLOW / Nullable ────────────────────────────────────

def compute_nullable(rules):
    """Compute which nonterminals can derive ε."""
    nullable = set()
    changed = True
    while changed:
        changed = False
        for name, body in rules.items():
            if name not in nullable and is_nullable(body, nullable):
                nullable.add(name)
                changed = True
    return nullable


def is_nullable(node, nullable):
    """Check if a grammar node can derive ε."""
    if node == EPSILON:
        return True
    if isinstance(node, Terminal):
        return False
    if isinstance(node, Nonterminal):
        return node.name in nullable
    if isinstance(node, Sequence):
        return all(is_nullable(item, nullable) for item in node.items)
    if isinstance(node, Alternation):
        return any(is_nullable(alt, nullable) for alt in node.alts)
    if isinstance(node, Optional):
        return True
    if isinstance(node, Repetition):
        return True
    return False


def compute_first(node, rules, nullable, memo=None):
    """Compute the FIRST set of a grammar node (set of terminals)."""
    if memo is None:
        memo = {}

    key = id(node)
    if key in memo:
        return memo[key]
    # Mark as being computed (avoid infinite recursion)
    memo[key] = set()
    result = set()

    if node == EPSILON:
        pass
    elif isinstance(node, Terminal):
        result.add(node)
    elif isinstance(node, Nonterminal):
        if node.name in rules:
            result |= compute_first(rules[node.name], rules, nullable, memo)
    elif isinstance(node, Sequence):
        for item in node.items:
            result |= compute_first(item, rules, nullable, memo)
            if not is_nullable(item, nullable):
                break
    elif isinstance(node, Alternation):
        for alt in node.alts:
            result |= compute_first(alt, rules, nullable, memo)
    elif isinstance(node, Optional):
        result |= compute_first(node.body, rules, nullable, memo)
    elif isinstance(node, Repetition):
        result |= compute_first(node.body, rules, nullable, memo)

    memo[key] = result
    return result


# ── LL(1) Conflict Detection ─────────────────────────────────────

def check_ll1(rules):
    """Check all alternations for LL(1) conflicts. Returns list of issues."""
    nullable = compute_nullable(rules)
    issues = []
    memo = {}

    def check_node(node, path):
        if isinstance(node, Alternation):
            # Check pairwise FIRST set disjointness
            alt_firsts = []
            for j, alt in enumerate(node.alts):
                first = compute_first(alt, rules, nullable, memo)
                alt_firsts.append((j, alt, first))

            for a_idx in range(len(alt_firsts)):
                for b_idx in range(a_idx + 1, len(alt_firsts)):
                    j_a, alt_a, first_a = alt_firsts[a_idx]
                    j_b, alt_b, first_b = alt_firsts[b_idx]
                    overlap = first_a & first_b
                    if overlap:
                        overlap_str = ", ".join(str(t) for t in sorted(overlap, key=str))
                        issues.append(
                            f"  {path}: alternatives {j_a+1} and {j_b+1} "
                            f"share FIRST tokens: {{{overlap_str}}}\n"
                            f"    alt {j_a+1}: {alt_a}\n"
                            f"    alt {j_b+1}: {alt_b}"
                        )

            # Recurse into each alternative
            for j, alt in enumerate(node.alts):
                check_node(alt, f"{path}|{j+1}")

        elif isinstance(node, Sequence):
            for j, item in enumerate(node.items):
                check_node(item, f"{path}.{j+1}")

        elif isinstance(node, Optional):
            # Optional: FIRST(body) must not overlap with FOLLOW
            # (We approximate: just recurse into body)
            check_node(node.body, f"{path}[?]")

        elif isinstance(node, Repetition):
            check_node(node.body, f"{path}{{*}}")

    for name, body in rules.items():
        check_node(body, name)

    return issues


# ── Main ─────────────────────────────────────────────────────────

def main():
    grammar_path = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("grammar/concrete.ebnf")
    if not grammar_path.exists():
        print(f"error: grammar file not found: {grammar_path}", file=sys.stderr)
        sys.exit(1)

    text = grammar_path.read_text()
    try:
        rules = parse_ebnf(text)
    except ValueError as e:
        print(f"error: failed to parse grammar: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Parsed {len(rules)} grammar rules from {grammar_path}")

    issues = check_ll1(rules)

    if issues:
        print(f"\nFound {len(issues)} LL(1) conflict(s):\n")
        for issue in issues:
            print(issue)
            print()
        sys.exit(1)
    else:
        print("LL(1) check passed: no FIRST/FIRST conflicts found.")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/check_ll1.rs
///! LL(1) grammar checker for Concrete (Rust implementation)
///!
///! Parses grammar/concrete.ebnf and verifies the LL(1) property:
///! for every alternation, FIRST sets of alternatives must be disjoint.
///!
///! Build:  rustc -O -o scripts/check_ll1_rs scripts/check_ll1.rs
///! Usage:  ./scripts/check_ll1_rs [grammar/concrete.ebnf]
///!
///! Exit 0 = LL(1) clean, 1 = conflicts found.

use std::collections::{HashMap, HashSet};
use std::env;
use std::fs;
use std::process;

// ── AST ──────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
enum Node {
    Terminal(String),
    Nonterminal(String),
    Sequence(Vec<Node>),
    Alternation(Vec<Node>),
    Optional(Box<Node>),
    Repetition(Box<Node>),
    Epsilon,
}

// ── EBNF Tokenizer ──────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq)]
enum Tok {
    Term(String),
    Id(String),
    Eq,
    Semi,
    Pipe,
    LBrack,
    RBrack,
    LBrace,
    RBrace,
    LParen,
    RParen,
    Comma,
    Plus,
    Dot,
    End,
}

const TOKEN_CLASSES: &[&str] = &["IDENT", "INT", "FLOAT", "STRING", "CHAR", "LABEL"];

fn is_token_class(s: &str) -> bool {
    TOKEN_CLASSES.contains(&s)
}

fn tokenize(text: &str) -> Vec<Tok> {
    let bytes = text.as_bytes();
    let mut toks = Vec::new();
    let mut i = 0;
    while i < bytes.len() {
        let ch = bytes[i] as char;
        // Skip whitespace
        if ch.is_ascii_whitespace() {
            i += 1;
            continue;
        }
        // Skip comments (* ... *)
        if i + 1 < bytes.len() && bytes[i] == b'(' && bytes[i + 1] == b'*' {
            match text[i + 2..].find("*)") {
                Some(pos) => {
                    i = i + 2 + pos + 2;
                    continue;
                }
                None => {
                    eprintln!("error: unterminated comment at position {i}");
                    process::exit(2);
                }
            }
        }
        // Single-char tokens
        let tok = match ch {
            '=' => Some(Tok::Eq),
            ';' => Some(Tok::Semi),
            '|' => Some(Tok::Pipe),
            '[' => Some(Tok::LBrack),
            ']' => Some(Tok::RBrack),
            '{' => Some(Tok::LBrace),
            '}' => Some(Tok::RBrace),
            '(' => Some(Tok::LParen),
            ')' => Some(Tok::RParen),
            ',' => Some(Tok::Comma),
            '+' => Some(Tok::Plus),
            '.' => Some(Tok::Dot),
            _ => None,
        };
        if let Some(t) = tok {
            toks.push(t);
            i += 1;
            continue;
        }
        // Quoted terminal
        if ch == '\'' {
            let j = text[i + 1..]
                .find('\'')
                .unwrap_or_else(|| {
                    eprintln!("error: unterminated terminal at position {i}");
                    process::exit(2);
                })
                + i
                + 1;
            let val = format!("'{}'", &text[i + 1..j]);
            toks.push(Tok::Term(val));
            i = j + 1;
            continue;
        }
        // Identifier
        if ch.is_ascii_alphabetic() || ch == '_' {
            let j = text[i..]
                .find(|c: char| !c.is_ascii_alphanumeric() && c != '_')
                .map(|p| p + i)
                .unwrap_or(bytes.len());
            let val = &text[i..j];
            toks.push(Tok::Id(val.to_string()));
            i = j;
            continue;
        }
        eprintln!("error: unexpected character {ch:?} at position {i}");
        process::exit(2);
    }
    toks.push(Tok::End);
    toks
}

// ── EBNF Parser ─────────────────────────────────────────────────

struct Parser {
    toks: Vec<Tok>,
    pos: usize,
}

impl Parser {
    fn new(toks: Vec<Tok>) -> Self {
        Self { toks, pos: 0 }
    }

    fn peek(&self) -> &Tok {
        &self.toks[self.pos]
    }

    fn advance(&mut self) -> Tok {
        let t = self.toks[self.pos].clone();
        self.pos += 1;
        t
    }

    fn expect(&mut self, expected: &Tok) {
        let got = self.advance();
        if std::mem::discriminant(&got) != std::mem::discriminant(expected) {
            eprintln!("error: expected {expected:?}, got {got:?} at token {}", self.pos - 1);
            process::exit(2);
        }
    }

    fn parse_alternation(&mut self) -> Node {
        let mut alts = vec![self.parse_sequence()];
        while *self.peek() == Tok::Pipe {
            self.advance();
            alts.push(self.parse_sequence());
        }
        if alts.len() == 1 {
            alts.pop().unwrap()
        } else {
            Node::Alternation(alts)
        }
    }

    fn parse_sequence(&mut self) -> Node {
        let mut items = Vec::new();
        loop {
            match self.peek() {
                Tok::Pipe | Tok::Semi | Tok::RBrack | Tok::RBrace | Tok::RParen | Tok::End => {
                    break
                }
                _ => items.push(self.parse_atom()),
            }
        }
        match items.len() {
            0 => Node::Epsilon,
            1 => items.pop().unwrap(),
            _ => Node::Sequence(items),
        }
    }

    fn parse_atom(&mut self) -> Node {
        match self.peek().clone() {
            Tok::LBrack => {
                self.advance();
                let body = self.parse_alternation();
                self.expect(&Tok::RBrack);
                Node::Optional(Box::new(body))
            }
            Tok::LBrace => {
                self.advance();
                let body = self.parse_alternation();
                self.expect(&Tok::RBrace);
                Node::Repetition(Box::new(body))
            }
            Tok::LParen => {
                self.advance();
                let body = self.parse_alternation();
                self.expect(&Tok::RParen);
                body
            }
            Tok::Term(val) => {
                self.advance();
                Node::Terminal(val)
            }
            Tok::Id(val) => {
                self.advance();
                if is_token_class(&val) {
                    Node::Terminal(val)
                } else {
                    Node::Nonterminal(val)
                }
            }
            other => {
                eprintln!("error: unexpected token {other:?} at position {}", self.pos);
                process::exit(2);
            }
        }
    }

    fn parse_rules(&mut self) -> Vec<(String, Node)> {
        let mut rules = Vec::new();
        while *self.peek() != Tok::End {
            let name = match self.advance() {
                Tok::Id(n) => n,
                other => {
                    eprintln!("error: expected rule name, got {other:?}");
                    process::exit(2);
                }
            };
            self.expect(&Tok::Eq);
            let body = self.parse_alternation();
            self.expect(&Tok::Semi);
            rules.push((name, body));
        }
        rules
    }
}

// ── Nullable ────────────────────────────────────────────────────

fn compute_nullable(rules: &[(String, Node)]) -> HashSet<String> {
    let mut nullable = HashSet::new();
    let mut changed = true;
    while changed {
        changed = false;
        for (name, body) in rules {
            if !nullable.contains(name) && is_nullable(body, &nullable) {
                nullable.insert(name.clone());
                changed = true;
            }
        }
    }
    nullable
}

fn is_nullable(node: &Node, nullable: &HashSet<String>) -> bool {
    match node {
        Node::Epsilon => true,
        Node::Terminal(_) => false,
        Node::Nonterminal(name) => nullable.contains(name),
        Node::Sequence(items) => items.iter().all(|i| is_nullable(i, nullable)),
        Node::Alternation(alts) => alts.iter().any(|a| is_nullable(a, nullable)),
        Node::Optional(_) | Node::Repetition(_) => true,
    }
}

// ── FIRST set ───────────────────────────────────────────────────

fn compute_first(
    node: &Node,
    rules: &HashMap<String, &Node>,
    nullable: &HashSet<String>,
    visiting: &mut HashSet<String>,
) -> HashSet<String> {
    match node {
        Node::Epsilon => HashSet::new(),
        Node::Terminal(name) => {
            let mut s = HashSet::new();
            s.insert(name.clone());
            s
        }
        Node::Nonterminal(name) => {
            if visiting.contains(name) {
                return HashSet::new(); // cycle guard
            }
            if let Some(body) = rules.get(name) {
                visiting.insert(name.clone());
                let result = compute_first(body, rules, nullable, visiting);
                visiting.remove(name);
                result
            } else {
                HashSet::new()
            }
        }
        Node::Sequence(items) => {
            let mut result = HashSet::new();
            for item in items {
                result.extend(compute_first(item, rules, nullable, visiting));
                if !is_nullable(item, nullable) {
                    break;
                }
            }
            result
        }
        Node::Alternation(alts) => {
            let mut result = HashSet::new();
            for alt in alts {
                result.extend(compute_first(alt, rules, nullable, visiting));
            }
            result
        }
        Node::Optional(body) | Node::Repetition(body) => {
            compute_first(body, rules, nullable, visiting)
        }
    }
}

// ── LL(1) check ─────────────────────────────────────────────────

fn check_node(
    node: &Node,
    path: &str,
    rules: &HashMap<String, &Node>,
    nullable: &HashSet<String>,
    issues: &mut Vec<String>,
) {
    match node {
        Node::Alternation(alts) => {
            let firsts: Vec<HashSet<String>> = alts
                .iter()
                .map(|alt| {
                    let mut visiting = HashSet::new();
                    compute_first(alt, rules, nullable, &mut visiting)
                })
                .collect();

            for a in 0..firsts.len() {
                for b in (a + 1)..firsts.len() {
                    let overlap: Vec<&String> =
                        firsts[a].intersection(&firsts[b]).collect();
                    if !overlap.is_empty() {
                        let mut sorted: Vec<&str> =
                            overlap.iter().map(|s| s.as_str()).collect();
                        sorted.sort();
                        issues.push(format!(
                            "  {path}: alternatives {} and {} share FIRST tokens: {{{}}}",
                            a + 1,
                            b + 1,
                            sorted.join(", ")
                        ));
                    }
                }
            }

            for (i, alt) in alts.iter().enumerate() {
                check_node(alt, &format!("{path}|{}", i + 1), rules, nullable, issues);
            }
        }
        Node::Sequence(items) => {
            for (i, item) in items.iter().enumerate() {
                check_node(item, &format!("{path}.{}", i + 1), rules, nullable, issues);
            }
        }
        Node::Optional(body) | Node::Repetition(body) => {
            check_node(body, path, rules, nullable, issues);
        }
        _ => {}
    }
}

// ── Main ────────────────────────────────────────────────────────

fn main() {
    let args: Vec<String> = env::args().collect();
    let path = if args.len() > 1 {
        &args[1]
    } else {
        "grammar/concrete.ebnf"
    };

    let text = match fs::read_to_string(path) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("error: cannot open {path}: {e}");
            process::exit(1);
        }
    };

    let toks = tokenize(&text);
    let mut parser = Parser::new(toks);
    let rule_list = parser.parse_rules();

    let nullable = compute_nullable(&rule_list);

    let rule_map: HashMap<String, &Node> =
        rule_list.iter().map(|(n, b)| (n.clone(), b)).collect();

    println!("Parsed {} grammar rules from {path}", rule_list.len());

    let mut issues = Vec::new();
    for (name, body) in &rule_list {
        check_node(body, name, &rule_map, &nullable, &mut issues);
    }

    if !issues.is_empty() {
        println!("\nFound {} LL(1) conflict(s):\n", issues.len());
        for issue in &issues {
            println!("{issue}\n");
        }
        process::exit(1);
    }

    println!("LL(1) check passed: no FIRST/FIRST conflicts found.");
}
